### PR TITLE
Add support for extracting comment replies continuation

### PR DESCRIPTION
**Pull request by FireMasterK**
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.
- [ ] I have tested the API against [NewPipe](https://github.com/TeamNewPipe/NewPipe).
- [ ] I agree to create a pull request for [NewPipe](https://github.com/TeamNewPipe/NewPipe) as soon as possible to make it compatible with the changed API.

This takes an alternative approach to #576

@Stypox could you verify that this approach is okay?

### CODE REVIEW

**Comment by Stypox** [comment]
Does it work also with comments continuations? Would this approach allow having both nested replies (for Peertube, I think) and replies continuations (when not all replies are loaded at once)?

**Comment by FireMasterK** [comment]
> Does it work also with comments continuations

Yes! Nothing should break it, although I will check it manually. I would have to check this once again as I have not check if there's a different JSON structure for these continuations.

> Would this approach allow having both nested replies

Yes, each comment would have an option replies `Page`. In YouTube there's only one level, so it would be null there.

To get reply Comments, you just call `getMoreItems` with the replies `Page`.


> replies continuations

This would be a regular continuation / getNextPage(), nothing should change.

**Comment by Stypox** [comment]
@FireMasterK ok, then this approach is good. Thank you :-)

**Comment by FireMasterK** [comment]
Just one small issue - this implementation doesn't support multi-level threads in one page request, but I think that can be done later. (I'm not sure if any services do this either)

**Comment by Mhowser** [comment]
Is this PR ready to be reviewed and merged soon?

File: extractor/src/main/java/org/schabi/newpipe/extractor/comments/CommentsInfoItemExtractor.java [hunk 1]
```diff
@@ -107,4 +108,13 @@ default boolean isUploaderVerified() throws ParsingException {
     default int getStreamPosition() throws ParsingException {
         return CommentsInfoItem.NO_STREAM_POSITION;
     }
+
+    /**
+     * The continuation page which is used to get comment replies from.
+     * @return the continuation Page for the replies
```
Comment (Stypox) [inline extractor/src/main/java/org/schabi/newpipe/extractor/comments/CommentsInfoItemExtractor.java:114]:
```suggestion
     * @return the continuation Page for the replies, or null if replies are not supported
```

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeCommentsExtractor.java [hunk 2]
```diff
@@ -212,10 +215,12 @@ private void collectCommentsFrom(final CommentsInfoItemsCollector collector,
             contents.remove(index);
         }
 
+        String jsonKey = contents.getObject(0).has("commentThreadRenderer") ? "commentThreadRenderer" : "commentRenderer";
```
Comment (Stypox) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeCommentsExtractor.java:218]:
```suggestion
        final String jsonKey = contents.getObject(0).has("commentThreadRenderer") ? "commentThreadRenderer" : "commentRenderer";
```

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeCommentsExtractor.java [hunk 3]
```diff
@@ -212,10 +215,12 @@ private void collectCommentsFrom(final CommentsInfoItemsCollector collector,
...
+
         final List<Object> comments;
         try {
             comments = JsonUtils.getValues(contents,
-                    "commentThreadRenderer.comment.commentRenderer");
+                    jsonKey);
```
Comment (Stypox) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeCommentsExtractor.java:223]: Put this on one line

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeCommentsInfoItemExtractor.java [hunk 4]
```diff
@@ -29,6 +32,16 @@ public YoutubeCommentsInfoItemExtractor(final JsonObject json,
         this.timeAgoParser = timeAgoParser;
     }
 
+    private JsonObject getCommentRenderer() throws ParsingException {
```
Comment (Stypox) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeCommentsInfoItemExtractor.java:35]: Why is there `throws ParsingException`?

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeCommentsInfoItemExtractor.java [hunk 5]
```diff
@@ -224,10 +237,20 @@ public String getUploaderName() throws ParsingException {
...
         }
     }
+
+    @Override
+    public Page getReplies() throws ParsingException {
+        try {
+            final String id = JsonUtils.getString(JsonUtils.getArray(json, "replies.commentRepliesRenderer.contents").getObject(0), "continuationItemRenderer.continuationEndpoint.continuationCommand.token");
+            return new Page(url, id);
+        } catch (final Exception e) {
+            return null;
```
Comment (Stypox) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeCommentsInfoItemExtractor.java:253]: I guess if `YoutubeCommentsInfoItemExtractor` is used as a reply extractor this `return null` would kick in for YouTube, wouldn't it? Add a comment to explain this.

**Review by Stypox — CHANGES_REQUESTED** [review]

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeCommentsInfoItemExtractor.java (hunk 4 above)
Comment (FireMasterK) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeCommentsInfoItemExtractor.java:35]: `JsonUtils` throws it, that's why.

Comment (Stypox) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeCommentsInfoItemExtractor.java:35]: Okk

**Review by Stypox — APPROVED** [review]
Perfect, thank you :-D

**Comment by Stypox** [comment]
This will be merged after release 0.20.10 is released

**Comment by TobiGr** [comment]
Do we need to squash the commits?

**Comment by Stypox** [comment]
I don't think that's needed

**Pull request merged**